file/random_access_file_reader.cc

```diff
@@ -457,9 +457,16 @@ IOStatus RandomAccessFileReader::ReadAsync(
 
   IOStatus s = file_->ReadAsync(req, opts, read_async_callback, read_async_info,
                                 io_handle, del_fn, nullptr /*dbg*/);
+// Suppress false positive clang analyzer warnings.
+// Memory is not released if file_->ReadAsync returns !s.ok(), because
+// ReadAsyncCallback is never called in that case. If ReadAsyncCallback is
+// called then ReadAsync should always return IOStatus::OK().
+#ifndef __clang_analyzer__
```

Comment:
@anand1976 I wanted to review this section. clang analyzer was giving false warning of double delete. I checked the flow also it provided in output and we never call callback if ReadAsync returns Not OK Status. Let me know if there is a better way to do this.

Reply:
I'll have to look at the clang analyzer output to be sure. I wonder if the issue is in PosixRandomAccessFile::ReadAsync. When it returns IOError (https://github.com/facebook/rocksdb/blob/main/env/io_posix.cc#L926), the posix_handle is not deleted and it has a reference to the callback. I don't know if clang analyzer can guess that the callback will never be called in that case.

Reply:
Let me rerun and share the clang analyzer result again with you with exact code path.

Reply:
Flow is:
      - calls RandomAccessFileReader::ReadAsync(
      - use_direct_io is false so it allocates ReadAsyncInfo object.
      - calls FSRandomAccessFile::ReadAsync
      - calls callback cb(req, cb_arg);
      - calls RandomAccessFileReader::ReadAsyncCallback
      - everything goes right then delete read_async_info in RandomAccessFileReader::ReadAsyncCallback
      - In RandomAccessFileReader::ReadAsync it takes false branch of (!s.ok()) and delete the read_async_info
      which is giving the double deletion error.

It's not calling the PosixRandomAccessFile::ReadAsync in the failure case.

Reply:
It gives the error of double deletion  of `read_async_info` in line 466.

Reply:
I don't see anything wrong in the flow you mentioned. I guess clang analyzer is wrong.

Reply:
Yes, clang analyzer goes through all branches but in our case that's not possible.